services/stats/stats.go

```diff
@@ -93,327 +78,190 @@ func (t Tags) String() string {
+	return c, err
 }
 
-func newTaggedStat(Name, StatType string, tags Tags, samplingRate float32) (rStats RudderStats) {
-	// If stats is not enabled, returning a dummy struct
-	if !statsEnabled {
-		return &RudderStatsT{
-			Name:        Name,
-			StatType:    StatType,
-			Client:      nil,
-			dontProcess: true,
-		}
-	}
-
-	// Clean up tags based on deployment type. No need to send workspace id tag for free tier customers.
-	tags = CleanupTagsBasedOnDeploymentType(tags, "workspaceId")
-
-	if tags == nil {
-		tags = make(Tags)
-	}
-	// key comprises of the measurement name plus all tag-value pairs
-	taggedClientKey := StatType + tags.String()
-
-	taggedClientsMapLock.RLock()
-	taggedClient, found := taggedClientsMap[taggedClientKey]
-	taggedClientsMapLock.RUnlock()
-
-	if !found {
-		taggedClientsMapLock.Lock()
-		if taggedClient, found = taggedClientsMap[taggedClientKey]; !found { // double check for race
-			tagVals := tags.Strings()
-			if !connEstablished {
-				taggedClientPendingTags = append(taggedClientPendingTags, taggedClientKey)
-				taggedClientPendingKeys = append(taggedClientPendingKeys, tagVals)
-			}
-			taggedClient = client.Clone(conn, statsd.TagsFormat(getTagsFormat()), defaultTags(), statsd.Tags(tagVals...), statsd.SampleRate(samplingRate))
-			taggedClientsMap[taggedClientKey] = taggedClient
+func (s *statsdStats) collectPeriodicStats() {
+	gaugeFunc := func(key string, val uint64) {
+		if s.state.client.statsd != nil {
+			s.state.client.statsd.Gauge("runtime_"+key, val)
 		}
-		taggedClientsMapLock.Unlock()
 	}
+	s.state.rc = newRuntimeStatsCollector(gaugeFunc)
+	s.state.rc.PauseDur = time.Duration(s.conf.periodic.statsCollectionInterval) * time.Second
+	s.state.rc.EnableCPU = s.conf.periodic.enableCPUStats
+	s.state.rc.EnableMem = s.conf.periodic.enableMemStats
+	s.state.rc.EnableGC = s.conf.periodic.enableGCStats
 
-	return &RudderStatsT{
-		Name:        Name,
-		StatType:    StatType,
-		Client:      taggedClient,
-		dontProcess: false,
-	}
-}
-
-func NewTaggedStat(Name, StatType string, tags Tags) (rStats RudderStats) {
-	return DefaultStats.NewTaggedStat(Name, StatType, tags)
-}
-
-func (rStats *RudderStatsT) Count(n int) {
-	if !statsEnabled || rStats.dontProcess {
-		return
-	}
-	if rStats.StatType != CountType {
-		panic(fmt.Errorf("rStats.StatType:%s is not count", rStats.StatType))
-	}
-	rStats.Client.Count(rStats.Name, n)
-}
-
-// Increment increases the stat by 1. Is the Equivalent of Count(1). Only applies to CountType stats
-func (rStats *RudderStatsT) Increment() {
-	if !statsEnabled || rStats.dontProcess {
-		return
-	}
-	if rStats.StatType != CountType {
-		panic(fmt.Errorf("rStats.StatType:%s is not count", rStats.StatType))
+	s.state.mc = newMetricStatsCollector(s, s.conf.periodic.metricManager)
+	if s.conf.periodic.enabled {
+		var wg sync.WaitGroup
+		wg.Add(2)
+		go func() {
+			defer wg.Done()
+			s.state.rc.run()
+		}()
+		go func() {
+			defer wg.Done()
+			s.state.mc.run()
```

Comment:
Is it safe to put metrics collector behind `s.conf.periodic.enabled` flag?

Reply:
Yes it was like that previously. The idea is that both routines collect periodical stats at the same frequency

Reply:
My only fear was if someone disables runtime metrics (not wanting cpu/memory usage metrics), pending_events metrics will go away. But if it was like that previously as well, we can leave this as it is.